{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iSNV diversity calculations for H7N3 avian influenza virus genomes \n",
    "This Juptyer notebook was used to calculate iSNV diversity for avian influenza virus genomes sequenced from turkeys and chickens infected with H7N3 low or high pathogenicity avian viruses from the 2020 outbreak in North and South Carolina, United States. Its main goal is to calculate iSNV diversity metrics namely Shannon entropy and nucleotide diversity from variant data as determined by Lofreq. Lofreq outputs vcf files that contains the nucleotide change, position, and frequency of an identified iSNV. The vcf files are then converted into a dataframe from which iSNV diversity metrics are calculated from.\n",
    "\n",
    "## Relevant references:\n",
    "\n",
    "#### 1. Lofreq\n",
    "##### Wilm A, Aw PP, Bertrand D, Yeo GH, Ong SH, Wong CH, Khor CC, Petric R, Hibberd ML, Nagarajan N. 2012. LoFreq: a sequence-quality aware, ultra-sensitive variant caller for uncovering cell-population heterogeneity from high-throughput sequencing datasets. Nucleic Acids Res 40:11189-201.\n",
    "\n",
    "#### 2. Shannon entropy and Frequency cutoff of 2%\n",
    "##### Fusaro A, Tassoni L, Milani A, Hughes J, Salviato A, Murcia PR, Massi P, Zamperin G, Bonfanti L, Marangon S, Cattoli G, Monne I. 2016. Unexpected Interfarm Transmission Dynamics during a Highly Pathogenic Avian Influenza Epidemic. J Virol 90:6401-11.\n",
    "##### McCrone JT, Lauring AS. 2016. Measurements of Intrahost Viral Diversity Are Extremely Sensitive to Systematic Errors in Variant Calling. J Virol 90:6884-95.\n",
    "\n",
    "#### 3. Nucleotide diversity\n",
    "##### Zhao L, Illingworth CJR. 2019. Measurements of intrahost viral diversity require an unbiased diversity metric. Virus Evol 5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, math, re\n",
    "\n",
    "wd='C:/Users/Christina.Leyson/OneDrive - USDA/Documents/Bioinformatics/1-H7N3 in turkeys and chickens/'\n",
    "\n",
    "path_input_sample_key=wd+'sample_key_onesheet 3-29-2021.xlsx'\n",
    "path_input_tables=wd+'MPJ-CL Plate 5.1-7.1/Lofreq tables/' #contains vcf files\n",
    "path_input_grpvirusdose=wd+'grp_virus_dose_key.xlsx'\n",
    "\n",
    "path_output_tables=wd+'MPJ-CL Plate 5.1-7.1/Lofreq tables/'\n",
    "path_output_diversity=wd+'MPJ-CL Plate 5.1-7.1/Lofreq summary ntdiversity/iSNV_diversity.xlsx'\n",
    "path_ouput_change_list=wd+'MPJ-CL Plate 5.1-7.1/Lofreq summary ntdiversity/SNV_list.xlsx'\n",
    "\n",
    "cutoff=0.02\n",
    "genome_length_dict={'HP':13608,'LP1':13581,'LP2':13515}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Create a sample key with information on the samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a function to define sample as either a swab or tissue sample\n",
    "def smpl_class(row): \n",
    "    if row['sample_type'] in ['OP','CL']:\n",
    "        return 'swab'\n",
    "    elif row['sample_type'] in ['BR','HR','LU','MS','SP']:\n",
    "        return 'tissue'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Unnamed: 0                                             int64\n",
       "SampleID                                              object\n",
       "R1                                                     int64\n",
       "R2                                                     int64\n",
       "Pear                                                   int64\n",
       "bam_wgs                                                int64\n",
       "bam_wgs_fasta                                          int64\n",
       "loreq                                                  int64\n",
       "Library prep#                                        float64\n",
       "Library plate                                        float64\n",
       "Library prep position                                 object\n",
       "RT-PCR date                                   datetime64[ns]\n",
       "RT-PCR#                                              float64\n",
       "RT-PCR Format                                         object\n",
       "RT-PCR plate/strip#                                  float64\n",
       "RT-PCR position                                       object\n",
       "Opti-RT-PCR result                                    object\n",
       "notes                                                 object\n",
       "RT-PCR type                                           object\n",
       "Species_code                                          object\n",
       "Group                                                float64\n",
       "Bird number                                          float64\n",
       "Exposure                                              object\n",
       "time_pt                                              float64\n",
       "sample_type                                           object\n",
       "Sample_code                                           object\n",
       "PB1, PB2, PA                                          object\n",
       "HA, NA, NS                                            object\n",
       "NP and M                                              object\n",
       "Experiment no.                                        object\n",
       "Purified PCR product concentration (ng/ul)           float64\n",
       "SampleID_firstrun                                     object\n",
       "Prep #                                               float64\n",
       "Virus                                                 object\n",
       "Species_name                                          object\n",
       "Plate_qRTPCR                                         float64\n",
       "Well_qRTPCR                                           object\n",
       "Plate-Well                                            object\n",
       "Task                                                  object\n",
       "Cт                                                   float64\n",
       "Quantity                                             float64\n",
       "log_Quantity                                         float64\n",
       "sample_class                                          object\n",
       "Dose                                                 float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_key=pd.read_excel(path_input_sample_key)\n",
    "sample_key['sample_class']=sample_key.apply(smpl_class, axis=1)\n",
    "\n",
    "grpvirusdose_df=pd.read_excel(path_input_grpvirusdose)\n",
    "sample_key['Dose']=sample_key['Group'].map(dict(zip(grpvirusdose_df['Group'],grpvirusdose_df['Dose'])))\n",
    "sample_key.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "virus_smplID_dict=pd.Series(sample_key.Virus.values,index=sample_key.SampleID).to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Reformat vcf file to tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vcf2excel(file,path_input):\n",
    "    data=pd.read_csv(path_input+file,sep='\\t',skiprows=18,\n",
    "                     names=['CHROM','Position','ID','Reference','Sample','QUAL','FILTER','INFO'])\n",
    "    INFO_list=data['INFO'].tolist()\n",
    "    dp=[]\n",
    "    af=[]\n",
    "    sb=[]\n",
    "    dp4=[]\n",
    "    indel=[]\n",
    "    hrun=[]\n",
    "    for info in INFO_list:\n",
    "        entry=info.split(';',6)\n",
    "        dp.append((entry[0]).split('=')[1])\n",
    "        af.append((entry[1]).split('=')[1])\n",
    "        sb.append((entry[2]).split('=')[1])\n",
    "        dp4.append((entry[3]).split('=')[1])\n",
    "        if len(entry)>4:\n",
    "            if entry[4] == 'INDEL':\n",
    "                indel.append(True)\n",
    "            else:\n",
    "                indel.append(False)\n",
    "    \n",
    "            if 'HRUN' in entry[5]:\n",
    "                hrun.append((entry[5]).split('=')[1])\n",
    "            else:\n",
    "                hrun.append('')\n",
    "        else: \n",
    "            indel.append('')\n",
    "            hrun.append('') \n",
    "    return data,dp,af,sb,dp4,indel,hrun"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>ID</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Sample</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>DP</th>\n",
       "      <th>AF</th>\n",
       "      <th>SB</th>\n",
       "      <th>DP4</th>\n",
       "      <th>INDEL</th>\n",
       "      <th>HRUN</th>\n",
       "      <th>Segment</th>\n",
       "      <th>Segment_no</th>\n",
       "      <th>Segment_name</th>\n",
       "      <th>Change_name</th>\n",
       "      <th>Segment_position</th>\n",
       "      <th>SNV_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LP1_8-NS</td>\n",
       "      <td>4</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1676</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=221;AF=0.348416;SB=1;DP4=62,82,36,41</td>\n",
       "      <td>221</td>\n",
       "      <td>0.348416</td>\n",
       "      <td>1</td>\n",
       "      <td>62,82,36,41</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8-NS</td>\n",
       "      <td>8</td>\n",
       "      <td>NS</td>\n",
       "      <td>8-NS_A4G</td>\n",
       "      <td>8-NS_4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LP1_8-NS</td>\n",
       "      <td>337</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1144</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=3747;AF=0.018148;SB=2;DP4=1796,1883,31,37</td>\n",
       "      <td>3747</td>\n",
       "      <td>0.018148</td>\n",
       "      <td>2</td>\n",
       "      <td>1796,1883,31,37</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8-NS</td>\n",
       "      <td>8</td>\n",
       "      <td>NS</td>\n",
       "      <td>8-NS_T337A</td>\n",
       "      <td>8-NS_337</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LP1_8-NS</td>\n",
       "      <td>447</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>6294</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=3911;AF=0.070570;SB=3;DP4=1776,1857,142,134</td>\n",
       "      <td>3911</td>\n",
       "      <td>0.070570</td>\n",
       "      <td>3</td>\n",
       "      <td>1776,1857,142,134</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8-NS</td>\n",
       "      <td>8</td>\n",
       "      <td>NS</td>\n",
       "      <td>8-NS_C447T</td>\n",
       "      <td>8-NS_447</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LP1_8-NS</td>\n",
       "      <td>611</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>90</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=3209;AF=0.002805;SB=4;DP4=1513,1687,6,3</td>\n",
       "      <td>3209</td>\n",
       "      <td>0.002805</td>\n",
       "      <td>4</td>\n",
       "      <td>1513,1687,6,3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8-NS</td>\n",
       "      <td>8</td>\n",
       "      <td>NS</td>\n",
       "      <td>8-NS_T611C</td>\n",
       "      <td>8-NS_611</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LP1_8-NS</td>\n",
       "      <td>622</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>62</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=3210;AF=0.002181;SB=1;DP4=1521,1681,4,3</td>\n",
       "      <td>3210</td>\n",
       "      <td>0.002181</td>\n",
       "      <td>1</td>\n",
       "      <td>1521,1681,4,3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8-NS</td>\n",
       "      <td>8</td>\n",
       "      <td>NS</td>\n",
       "      <td>8-NS_A622G</td>\n",
       "      <td>8-NS_622</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM  Position ID Reference Sample  QUAL FILTER  \\\n",
       "0  LP1_8-NS         4  .         A      G  1676   PASS   \n",
       "1  LP1_8-NS       337  .         T      A  1144   PASS   \n",
       "2  LP1_8-NS       447  .         C      T  6294   PASS   \n",
       "3  LP1_8-NS       611  .         T      C    90   PASS   \n",
       "4  LP1_8-NS       622  .         A      G    62   PASS   \n",
       "\n",
       "                                             INFO    DP        AF SB  \\\n",
       "0         DP=221;AF=0.348416;SB=1;DP4=62,82,36,41   221  0.348416  1   \n",
       "1    DP=3747;AF=0.018148;SB=2;DP4=1796,1883,31,37  3747  0.018148  2   \n",
       "2  DP=3911;AF=0.070570;SB=3;DP4=1776,1857,142,134  3911  0.070570  3   \n",
       "3      DP=3209;AF=0.002805;SB=4;DP4=1513,1687,6,3  3209  0.002805  4   \n",
       "4      DP=3210;AF=0.002181;SB=1;DP4=1521,1681,4,3  3210  0.002181  1   \n",
       "\n",
       "                 DP4 INDEL HRUN Segment Segment_no Segment_name Change_name  \\\n",
       "0        62,82,36,41               8-NS          8           NS    8-NS_A4G   \n",
       "1    1796,1883,31,37               8-NS          8           NS  8-NS_T337A   \n",
       "2  1776,1857,142,134               8-NS          8           NS  8-NS_C447T   \n",
       "3      1513,1687,6,3               8-NS          8           NS  8-NS_T611C   \n",
       "4      1521,1681,4,3               8-NS          8           NS  8-NS_A622G   \n",
       "\n",
       "  Segment_position  SNV_test  \n",
       "0           8-NS_4      True  \n",
       "1         8-NS_337      True  \n",
       "2         8-NS_447      True  \n",
       "3         8-NS_611      True  \n",
       "4         8-NS_622      True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Compile all data from vcf files into one dataframe\n",
    "plates=['5.1','6','7.1','7.1.1']\n",
    "for p in plates:\n",
    "    path_input=wd+f'MPJ-CL Plate {p} - Galaxy/Lofreq Plate {p}/'\n",
    "    file_list=os.listdir(path_input)\n",
    "    for file in file_list:\n",
    "        filename=(file.split('.'))[0]\n",
    "        data,dp,af,sb,dp4,indel,hrun=vcf2excel(file,path_input)\n",
    "        data['DP']=dp\n",
    "        data['AF']=af\n",
    "        data['SB']=sb\n",
    "        data['DP4']=dp4\n",
    "        data['INDEL']=indel\n",
    "        data['HRUN']=hrun\n",
    "        data['Segment']=data.CHROM.str.split('_').str[1]\n",
    "        data['Segment_no']=data.Segment.str.split('-').str[0]\n",
    "        data['Segment_name']=data.Segment.str.split('-').str[1]\n",
    "        data['Change_name']=data['Segment']+'_'+data['Reference']+data['Position'].astype(str)+data['Sample']\n",
    "        data['Segment_position']=data['Segment']+'_'+data['Position'].astype(str)\n",
    "        data['SNV_test']=np.where((data.Sample.str.fullmatch('[ATCG]'))&(data.Reference.str.fullmatch('[ATCG]')),True,False)\n",
    "        data.to_excel(path_output_tables+filename+'.xlsx')\n",
    "print('Done.')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Determine how many iSNVs are present in each sample, average iSNV frequency (also allelic frequency, AF), Shannon entropy, and nucleotide diversity. Also get a list of unique SNVs across the whole dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list_tables=os.listdir(path_input_tables)\n",
    "\n",
    "change_list=[]\n",
    "\n",
    "sampleID_list=[]\n",
    "meanAF=[]\n",
    "SNVcount=[]\n",
    "entropy=[]\n",
    "\n",
    "pi=[]\n",
    "\n",
    "for file in file_list_tables:\n",
    "    sampleID=file.split('_')[0]\n",
    "    data=pd.read_excel(path_input_tables+file)\n",
    "    \n",
    "    #Get all SNVs that are above 2%\n",
    "    data_cutoff=data.loc[(data['AF']>=cutoff)&(data['SNV_test']==True)]\n",
    "    \n",
    "    #Get a list of unique SNVs across the whole dataset\n",
    "    changes_in_sample=data_cutoff['Change_name'].tolist()\n",
    "    change_list=list(set(change_list+changes_in_sample))\n",
    "    \n",
    "    #Get mean allelic frequency AF and SNV count\n",
    "    sampleID_list.append(sampleID)\n",
    "    meanAF.append(data_cutoff['AF'].mean())\n",
    "    SNVcount.append(data_cutoff['Change_name'].count())\n",
    "    \n",
    "    #Get a list of all loci/position where SNVs were found\n",
    "    segpos_list=data_cutoff['Segment_position'].tolist()\n",
    "    #print(len(data['Segment_position'].tolist()),len(segpos_list))\n",
    "    \n",
    "    #Calculate H is the sum of p*log p, p=frequency of the allele\n",
    "    #See McCrone et al, JVI publication\n",
    "    H_list=[]\n",
    "    \n",
    "    #Calculate Dl which is 1-sum of p^2\n",
    "    #See Zhao and Illingworth, Virus Evolution publication\n",
    "    Dl_list=[]\n",
    "    \n",
    "    for locus in segpos_list:\n",
    "        locus_changes_df=data_cutoff.loc[data_cutoff['Segment_position']==locus]\n",
    "        AF_list=locus_changes_df['AF'].tolist()\n",
    "        \n",
    "        #Get the frequency of the reference allele but if AF=1, no reference AF is added\n",
    "        if 1 not in AF_list:     \n",
    "            ref_AF=1-(sum(AF_list))\n",
    "            AF_list.append(ref_AF)\n",
    "        \n",
    "        pisq_list=[]\n",
    "        for AF in AF_list:\n",
    "            #Calculated H for entropy\n",
    "            H=(AF*(math.log10(AF)))\n",
    "            H_list.append(H)\n",
    "           \n",
    "            pisq=AF**2\n",
    "            pisq_list.append(pisq)\n",
    "            #print(AF)\n",
    "        #print(pisq_list, sum(pisq_list))\n",
    "        \n",
    "        #Calculate Dl for nucleotide diversity\n",
    "        Dl=1-(sum(pisq_list))\n",
    "        #print(Dl)\n",
    "        Dl_list.append(Dl)\n",
    "        #print(pisq_list,Dl)\n",
    "    \n",
    "    \n",
    "    #Get virus genome length \n",
    "    virus=virus_smplID_dict.get(sampleID)\n",
    "    genome_length=genome_length_dict.get(virus,13581)\n",
    "    \n",
    "    #Calculate Shannon entropy in the sample\n",
    "    entropy_smpl=-(sum(H_list))/genome_length\n",
    "    entropy.append(entropy_smpl)\n",
    "\n",
    "    #Calculate nucleotide diversity pi in the sample\n",
    "    #print(Dl_list,genome_length)\n",
    "    pi_smpl=(sum(Dl_list))/genome_length\n",
    "    pi.append(pi_smpl)\n",
    "    \n",
    "#Check methods :->\n",
    "#print(sampleID,AF_list,data['Change_name'].count(),data_cutoff['Change_name'].count(),len(H_list),len(entropy),len(file_list_tables),len(pi))\n",
    "#print(len(segpos_list),len(Dl_list)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "diversity_calculations=pd.DataFrame({'SampleID':sampleID_list,\n",
    "                                  'mean_AF':meanAF,\n",
    "                                  'SNV_count':SNVcount,\n",
    "                                  'Shannon_entropy':entropy,\n",
    "                                  'Nucleotide_diversity':pi})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There a total of 4656 number of unique sites\n"
     ]
    }
   ],
   "source": [
    "print('There a total of '+str(len(change_list))+' number of unique sites') \n",
    "change_list_df=pd.DataFrame({'Change_name':change_list})\n",
    "change_list_df.to_excel(path_ouput_change_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Combine sample information and iSNV diversity calculations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Unnamed: 0                                             int64\n",
       "SampleID                                              object\n",
       "R1                                                     int64\n",
       "R2                                                     int64\n",
       "Pear                                                   int64\n",
       "bam_wgs                                                int64\n",
       "bam_wgs_fasta                                          int64\n",
       "loreq                                                  int64\n",
       "Library prep#                                        float64\n",
       "Library plate                                        float64\n",
       "Library prep position                                 object\n",
       "RT-PCR date                                   datetime64[ns]\n",
       "RT-PCR#                                              float64\n",
       "RT-PCR Format                                         object\n",
       "RT-PCR plate/strip#                                  float64\n",
       "RT-PCR position                                       object\n",
       "Opti-RT-PCR result                                    object\n",
       "notes                                                 object\n",
       "RT-PCR type                                           object\n",
       "Species_code                                          object\n",
       "Group                                                float64\n",
       "Bird number                                          float64\n",
       "Exposure                                              object\n",
       "time_pt                                              float64\n",
       "sample_type                                           object\n",
       "Sample_code                                           object\n",
       "PB1, PB2, PA                                          object\n",
       "HA, NA, NS                                            object\n",
       "NP and M                                              object\n",
       "Experiment no.                                        object\n",
       "Purified PCR product concentration (ng/ul)           float64\n",
       "SampleID_firstrun                                     object\n",
       "Prep #                                               float64\n",
       "Virus                                                 object\n",
       "Species_name                                          object\n",
       "Plate_qRTPCR                                         float64\n",
       "Well_qRTPCR                                           object\n",
       "Plate-Well                                            object\n",
       "Task                                                  object\n",
       "Cт                                                   float64\n",
       "Quantity                                             float64\n",
       "log_Quantity                                         float64\n",
       "sample_class                                          object\n",
       "Dose                                                 float64\n",
       "mean_AF                                              float64\n",
       "SNV_count                                              int64\n",
       "Shannon_entropy                                      float64\n",
       "Nucleotide_diversity                                 float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Combine calculations and sample information\n",
    "diversity_summary=pd.merge(sample_key,diversity_calculations,how='left',on='SampleID')\n",
    "diversity_summary.to_excel(path_output_diversity)\n",
    "diversity_summary.dtypes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
